vnext/Microsoft.ReactNative.Cxx/JSValue.h

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#ifndef MICROSOFT_REACTNATIVE_JSVALUE
+#define MICROSOFT_REACTNATIVE_JSVALUE
+
+#include "Crash.h"
+#include "winrt/Microsoft.ReactNative.Bridge.h"
+
+namespace winrt::Microsoft::ReactNative::Bridge {
+
+// Forward declarations
+struct JSValue;
+IJSValueReader MakeJSValueTreeReader(const JSValue &root) noexcept;
+IJSValueReader MakeJSValueTreeReader(JSValue &&root) noexcept;
+IJSValueWriter MakeJSValueTreeWriter(JSValue &resultValue) noexcept;
+
+// Type alias for JSValue object type
+using JSValueObject = std::map<std::string, JSValue, std::less<>>;
+
+// Type alias for JSValue array type
+using JSValueArray = std::vector<JSValue>;
+
+// A view for JSValueObject
+struct JSValueObjectView {
+  JSValueObjectView(const JSValueObject &object) noexcept;
+  auto begin() const noexcept;
+  auto end() const noexcept;
+
+ private:
+  const JSValueObject &m_object;
+};
+
+// A view for JSObjectArray
+struct JSValueArrayView {
+  JSValueArrayView(const JSValueArray &array) noexcept;
+  auto begin() const noexcept;
+  auto end() const noexcept;
+
+ private:
+  const JSValueArray &m_array;
+};
+
+// JSValue represents an immutable JavaScript value that can be passed as a parameter.
+// It is created to simplify working with IJSValueReader in some complex cases.
+// It takes more resources than direct use of IJSValueReader, but provides more flexibility.
+// The JSValue is an immutable and is safe to be used from multiple threads.
+// It is move-only to avoid unnecessary or unexpected copying of values.
+struct JSValue {
+  static const JSValue Null;
+  static const JSValueObject EmptyObject;
+  static const JSValueArray EmptyArray;
+  static const std::string EmptyString;
+
+  JSValue() noexcept;
+  JSValue(std::nullptr_t) noexcept;
+  JSValue(JSValueObject &&value) noexcept;
+  JSValue(JSValueArray &&value) noexcept;
+  JSValue(std::string &&value) noexcept;
+  template <class TStringView, std::enable_if_t<std::is_convertible_v<TStringView, std::string_view>, int> = 1>
+  JSValue(TStringView value) noexcept;
+  template <class TBool, std::enable_if_t<std::is_same_v<TBool, bool>, int> = 1>
+  JSValue(TBool value) noexcept;
+  template <class TInt, std::enable_if_t<std::is_integral_v<TInt> && !std::is_same_v<TInt, bool>, int> = 1>
+  JSValue(TInt value) noexcept;
+  JSValue(double value) noexcept;
+
+  JSValue(const JSValue &other) = delete;
+  JSValue &operator=(const JSValue &other) = delete;
+
+  JSValue(JSValue &&other) noexcept;
+  JSValue &operator=(JSValue &&other) noexcept;
+
+  ~JSValue() noexcept;
+
+  JSValue Copy() const noexcept;
+  static JSValueObject CopyObject(const JSValueObject &other) noexcept;
+  static JSValueArray CopyArray(const JSValueArray &other) noexcept;
+
+  JSValueType Type() const noexcept;
+  bool IsNull() const noexcept;
+  const JSValueObject &Object() const noexcept;
```

Comment:
>&Object() [](start = 22, length = 9)

This looks strange, but I guess it's required by formatting?

Reply:
Same for the custom dtor stuff we need to do

---
In reply to: [356910886](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356910886) [](ancestors = 356910886)

Reply:
Yes, this is the formatting adopted by RNW project.
@NickGerleman, how is it related to the custom dtor?

---
In reply to: [356917132](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356917132) [](ancestors = 356917132,356910886)

Reply:
Whoops, I added that to the wrong thread, That was meant to be on the std::variant comment

---
In reply to: [356918726](https://github.com/microsoft/react-native-windows/pull/3760#discussion_r356918726) [](ancestors = 356918726,356917132,356910886)